Mathlib/Topology/Homotopy/HomotopyGroup.lean

```diff
@@ -4,283 +4,583 @@ Released under Apache 2.0 license as described in the file LICENSE.
-  rfl
+/-- The `n`-dimensional generalized loops based at `x` in a space `X` are
+  continuous functions `I^n → X` that sends the boundary to `x`.
+  We allow an arbitrary indexing type `N` in place of `Fin n` here. -/
+def GenLoop : Set C(I^N, X) :=
+  {p | ∀ y ∈ Cube.boundary N, p y = x}
+#align gen_loop GenLoop
 
-end Cube
+-- mathport name: «exprΩ^»
+scoped[Topology] notation "Ω^" => GenLoop
 
-/--
-The `n`-dimensional generalized loops; functions `I^n → X` that send the boundary to the base point.
--/
-structure GenLoop (n : ℕ) (x : X) extends C(I^ n, X) where
-  boundary : ∀ y ∈ Cube.boundary n, toFun y = x
-#align gen_loop GenLoop
+variable {N X x}
 
 namespace GenLoop
 
-instance funLike : FunLike (GenLoop n x) (I^ n) fun _ => X where
+instance funLike : FunLike (Ω^ N X x) (I^N) fun _ => X
+    where
   coe f := f.1
   coe_injective' := fun ⟨⟨f, _⟩, _⟩ ⟨⟨g, _⟩, _⟩ _ => by congr
 #align gen_loop.fun_like GenLoop.funLike
 
 @[ext]
-theorem ext (f g : GenLoop n x) (H : ∀ y, f y = g y) : f = g :=
-  FunLike.ext f g H
+theorem ext (f g : Ω^ N X x) (H : ∀ y, f y = g y) : f = g :=
+  FunLike.coe_injective' (funext H)
 #align gen_loop.ext GenLoop.ext
 
 @[simp]
-theorem mk_apply (f : C(I^ n, X)) (H y) : (⟨f, H⟩ : GenLoop n x) y = f y :=
+theorem mk_apply (f : C(I^N, X)) (H y) : (⟨f, H⟩ : Ω^ N X x) y = f y :=
   rfl
 #align gen_loop.mk_apply GenLoop.mk_apply
 
-@[simp] -- porting note: new lemma
-theorem toContinuousMap_apply (f : GenLoop n x) (y) : f.toContinuousMap y = f y := rfl
+/-- Copy of a `GenLoop` with a new map from the unit cube equal to the old one.
+  Useful to fix definitional equalities. -/
+def copy (f : Ω^ N X x) (g : (I^N) → X) (h : g = f) : Ω^ N X x :=
+  ⟨⟨g, h.symm ▸ f.1.2⟩, by convert f.2⟩
+#align gen_loop.copy GenLoop.copy
 
-initialize_simps_projections GenLoop (toFun → apply)
+/- porting note: this now requires the `funLike` instance,
+  so the instance is now put before `copy`. -/
+theorem coe_copy (f : Ω^ N X x) {g : (I^N) → X} (h : g = f) : ⇑(copy f g h) = g :=
+  rfl
+#align gen_loop.coe_copy GenLoop.coe_copy
 
-/-- The constant `GenLoop` at `x`.
--/
-@[simps!] -- porting note: new attr
-def const : GenLoop n x :=
+theorem copy_eq (f : Ω^ N X x) {g : (I^N) → X} (h : g = f) : copy f g h = f := by
+  ext x
+  exact congr_fun h x
+#align gen_loop.copy_eq GenLoop.copy_eq
+
+theorem boundary (f : Ω^ N X x) : ∀ y ∈ Cube.boundary N, f y = x :=
+  f.2
+#align gen_loop.boundary GenLoop.boundary
+
+/-- The constant `GenLoop` at `x`. -/
+def const : Ω^ N X x :=
   ⟨ContinuousMap.const _ x, fun _ _ => rfl⟩
 #align gen_loop.const GenLoop.const
 
-instance inhabited : Inhabited (GenLoop n x) where default := const
-#align gen_loop.inhabited GenLoop.inhabited
+@[simp]
+theorem const_apply {t} : (@const N X _ x) t = x :=
+  rfl
+#align gen_loop.const_apply GenLoop.const_apply
 
-/-- Restrict a `GenLoop n x` with `n ≠ 0` to the diagonal of the cube. -/
-@[simps!]
-def diagonal [NeZero n] (f : GenLoop n x) : Path x x where
```

Comment:
Was this diagonal stuff new in mathlib4?

Reply:
`diagonal` is essentially the forward direction of [`genLoopEquivOfUnique`](https://github.com/leanprover-community/mathlib4/pull/4939/files#diff-07c8e6b42dce840525e2eb9f2d80957e624355af85a208b09eff59eda02dfcf7R465) which holds for any type while the backward direction only exists for `inhabited` types.

It seems that @urkud introduced it to port `genLoopOneEquivPathSelf` and [`pi1EquivFundamentalGroup`](https://github.com/leanprover-community/mathlib4/pull/4939/files#diff-07c8e6b42dce840525e2eb9f2d80957e624355af85a208b09eff59eda02dfcf7L263-L286), but I no longer needs it. However, it may be beneficial to keep (generalized versions) this def and [this theorem](https://github.com/leanprover-community/mathlib4/pull/4939/files#diff-07c8e6b42dce840525e2eb9f2d80957e624355af85a208b09eff59eda02dfcf7L196) and refactor `genLoopEquivOfUnique` to use it.

Reply:
Okay, but that can happen after the port, I think.